### PR TITLE
test(storage): benchmark supports ranged reads

### PR DESCRIPTION
**Pull request by coryan**
A lot of analytical workloads use "ranged" reads, i.e., they read a
portion of a larger object. There seems to be performance advantages
when reading up to 2MiB. Likewise, there are measurable performance
differences when the offset is on a 2MiB boundary and/or the read
size is on a 2 MiB boundary.

We would like to measure these effects, and therefore we need new
parameters for the benchmark.

By now, we have 4 of these "quantized" size ranges, I did some
refactoring to avoid code duplication.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9378)

### CODE REVIEW

File: google/cloud/storage/benchmarks/storage_throughput_vs_cpu_benchmark.cc [hunk 1]
```diff
@@ -331,18 +346,48 @@ void RunThread(ThroughputOptions const& options, std::string const& bucket_name,
...
+    if (quantum != 0) {
+      distribution = std::uniform_int_distribution<decltype(quantum)>(
+          minimum / quantum, maximum / quantum);
+    }
+    return [d = std::move(distribution), quantum](auto& g) mutable {
+      return quantum * d(g);
+    };
+  };
+
+  auto write_buffer_size_generator = quantized_range_generator(
```
Comment (dbolduc) [inline google/cloud/storage/benchmarks/storage_throughput_vs_cpu_benchmark.cc:362]: in a future PR, you might want to group quantized options in a struct. Then have member functions for things like `output_quantized_*`, `quantized_range_generator`, parsing values from the command line.

File: google/cloud/storage/benchmarks/storage_throughput_vs_cpu_benchmark.cc [hunk 2]
```diff
@@ -331,18 +346,48 @@ void RunThread(ThroughputOptions const& options, std::string const& bucket_name,
...
-  std::uniform_int_distribution<std::size_t> write_buffer_size_generator(
-      options.minimum_write_buffer_size / options.write_buffer_quantum,
-      options.maximum_write_buffer_size / options.write_buffer_quantum);
-  std::uniform_int_distribution<std::size_t> read_buffer_size_generator(
-      options.minimum_read_buffer_size / options.read_buffer_quantum,
-      options.maximum_read_buffer_size / options.read_buffer_quantum);
+
+  auto quantized_range_generator = [](auto minimum, auto maximum,
+                                      auto quantum) {
+    auto distribution = std::uniform_int_distribution<decltype(quantum)>(0, 0);
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/benchmarks/storage_throughput_vs_cpu_benchmark.cc]: Fixed.

File: google/cloud/storage/benchmarks/throughput_options.cc [hunk 3]
```diff
@@ -20,6 +20,28 @@
...
+Status ValidateQuantizedRange(std::string const& name, std::int64_t minimum,
+                              std::int64_t maximum, std::int64_t quantum) {
+  using ::google::cloud::StatusCode;
+  if (minimum > maximum) {
+    std::ostringstream os;
+    os << "Invalid range for " << name << " [" << minimum << ',' << maximum
+       << "]";
+    return google::cloud::Status{StatusCode::kInvalidArgument, os.str()};
+  }
+  if (quantum < 0 || (quantum > minimum && minimum != 0)) {
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/benchmarks/throughput_options.cc]: Done.

File: google/cloud/storage/benchmarks/storage_throughput_vs_cpu_benchmark.cc [hunk 4]
```diff
@@ -149,52 +150,66 @@ int main(int argc, char* argv[]) {
...
+      << absl::FormatDuration(absl::FromChrono(options->transfer_stall_timeout))
+      << "\n# Download Stall Timeout: "
+      << absl::FormatDuration(absl::FromChrono(options->download_stall_timeout))
+      << "\n# Minimum Sample Delay: "
+      << absl::FormatDuration(absl::FromChrono(options->minimum_sample_delay));
+
+  output_quantized_range("Read Offset", options->minimum_read_offset,
+                         options->maximum_read_offset,
+                         options->read_offset_quantum);
+  output_quantized_range("Read Size", options->minimum_read_size,
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/benchmarks/storage_throughput_vs_cpu_benchmark.cc:208]: Fixed

**Pull request merged**